MSVC compiler warnings regarding for_each_pair_early

@rprospero MSVC throws up a couple of additional compiler warnings regarding `for_each_pair_early` - thought you might want to take a look and see if anything is worth tweaking before this PR goes in.

```
In file included from ../src/classes/partialset.cpp:11:
../src/classes/../templates/algorithms.h:89:22: warning: comparison of two values with different enumeration types in switch statement ('EarlyReturn<bool>::Type' and 'EarlyReturn<std::__1::optional<bool> >::Type') [-Wenum-compare-switch]
                case EarlyReturn<decltype(lambda(0, 0).value())>::Continue:
                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/classes/partialset.cpp:264:5: note: in instantiation of function template specialization 'for_each_pair_early<(lambda at ../src/classes/partialset.cpp:264:49)>' requested here
    for_each_pair_early(0, atomTypes_.nItems(), [&](int typeI, int typeJ) -> EarlyReturn<bool> {
    ^
In file included from ../src/classes/partialset.cpp:11:
../src/classes/../templates/algorithms.h:87:22: warning: comparison of two values with different enumeration types in switch statement ('EarlyReturn<bool>::Type' and 'EarlyReturn<std::__1::optional<bool> >::Type') [-Wenum-compare-switch]
                case EarlyReturn<decltype(lambda(0, 0).value())>::Break:
                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/classes/../templates/algorithms.h:85:22: warning: comparison of two values with different enumeration types in switch statement ('EarlyReturn<bool>::Type' and 'EarlyReturn<std::__1::optional<bool> >::Type') [-Wenum-compare-switch]
                case EarlyReturn<decltype(lambda(0, 0).value())>::Return:
                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/classes/../templates/algorithms.h:89:22: warning: comparison of two values with different enumeration types in switch statement ('EarlyReturn<bool>::Type' and 'EarlyReturn<std::__1::optional<bool> >::Type') [-Wenum-compare-switch]
                case EarlyReturn<decltype(lambda(0, 0).value())>::Continue:
                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/classes/partialset.cpp:585:5: note: in instantiation of function template specialization 'for_each_pair_early<(lambda at ../src/classes/partialset.cpp:585:36)>' requested here
    for_each_pair_early(0, nTypes, [&](int typeI, int typeJ) -> EarlyReturn<bool> {
    ^
In file included from ../src/classes/partialset.cpp:11:
../src/classes/../templates/algorithms.h:87:22: warning: comparison of two values with different enumeration types in switch statement ('EarlyReturn<bool>::Type' and 'EarlyReturn<std::__1::optional<bool> >::Type') [-Wenum-compare-switch]
                case EarlyReturn<decltype(lambda(0, 0).value())>::Break:
                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/classes/../templates/algorithms.h:85:22: warning: comparison of two values with different enumeration types in switch statement ('EarlyReturn<bool>::Type' and 'EarlyReturn<std::__1::optional<bool> >::Type') [-Wenum-compare-switch]
                case EarlyReturn<decltype(lambda(0, 0).value())>::Return:
                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

_Originally posted by @trisyoungs in https://github.com/projectdissolve/dissolve/pull/414#issuecomment-720481435_

--- a/src/templates/algorithms.h
+++ b/src/templates/algorithms.h
@@ -42,6 +42,7 @@ template <typename T> class EarlyReturn
     std::optional<T> value_;
 
     public:
+    using inner = T;
     EarlyReturn(Type t = Type::Continue, std::optional<T> v = std::nullopt) : type_(t), value_(v){};
     EarlyReturn(const T &val) : type_(Type::Return), value_(val){};
     Type type() const { return type_; }
@@ -61,11 +62,11 @@ auto for_each_pair_early(Iter begin, Iter end, Lam lambda) -> decltype(lambda(0,
             auto result = lambda(i, *elem1, j, *elem2);
             switch (result.type())
             {
-                case EarlyReturn<decltype(lambda(0, *begin, 0, *end).value())>::Return:
+                case EarlyReturn<typename decltype(result)::inner>::Return:
                     return result.value();
-                case EarlyReturn<decltype(lambda(0, *begin, 0, *end).value())>::Break:
+                case EarlyReturn<typename decltype(result)::inner>::Break:
                     break;
-                case EarlyReturn<decltype(lambda(0, *begin, 0, *end).value())>::Continue:
+                case EarlyReturn<typename decltype(result)::inner>::Continue:
                     continue;
             }
         }
@@ -82,11 +83,11 @@ template <class Lam> auto for_each_pair_early(int begin, int end, Lam lambda) ->
             auto result = lambda(i, j);
             switch (result.type())
             {
-                case EarlyReturn<decltype(lambda(0, 0).value())>::Return:
+                case EarlyReturn<typename decltype(result)::inner>::Return:
                     return result.value();
-                case EarlyReturn<decltype(lambda(0, 0).value())>::Break:
+                case EarlyReturn<typename decltype(result)::inner>::Break:
                     break;
-                case EarlyReturn<decltype(lambda(0, 0).value())>::Continue:
+                case EarlyReturn<typename decltype(result)::inner>::Continue:
                     continue;
             }
         }

